### PR TITLE
Performance improvement for the connection phase

### PR DESCRIPTION
**Pull request by maciejzaleski**
Improved performance of the VerifyPoupardStern by verifying  x values in parallel rather than sequentially.

### CODE REVIEW

File: NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs [hunk 1]
```diff
@@ -160,17 +160,23 @@ public static bool VerifyPoupardStern(this RsaKeyParameters pubKey, PoupardStern
             rPrime = y.Subtract(Modulus.Multiply(w));
 
             // Verifying x values
+            Dictionary<int, bool> numbers = new Dictionary<int, bool>();
```
Comment (NicolasDorier) [inline NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs]: Use array `var numbers = new bool[BigK] ;`

File: NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs [hunk 2]
```diff
@@ -160,17 +160,23 @@ public static bool VerifyPoupardStern(this RsaKeyParameters pubKey, PoupardStern
...
                 var rs = z_i.ModPow(rPrime, Modulus);
                 // If the two sides are not equal
                 if (!(proof.XValues[i].Equals(rs)))
-                    return false;
-            }
+                    numbers[i] = false;
+            });
 
-            return true;
+            return !numbers.ContainsValue(false);
```
Comment (NicolasDorier) [inline NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs]: `return numbers.All();` (or `return numbers.All(o => o);`)

File: NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs [hunk 3]
```diff
@@ -160,17 +160,23 @@ public static bool VerifyPoupardStern(this RsaKeyParameters pubKey, PoupardStern
             rPrime = y.Subtract(Modulus.Multiply(w));
 
             // Verifying x values
+            Dictionary<int, bool> numbers = new Dictionary<int, bool>();
             for (int i = 0; i < BigK; i++)
+                numbers.Add(i, true);
+
+            Parallel.ForEach(numbers, (n) =>
```
Comment (NicolasDorier) [inline NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs]: `Parallel.ForEach(Enumerable.Range(0, numbers.Length), (n) =>`

File: NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs (hunk 1 above)
Comment (maciejzaleski) [reply NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs]: Changed to bool array with a comment that all elements in the array will get initialised to false. The logic within the loop has now changed to set array elements to `true`.

File: NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs (hunk 2 above)
Comment (maciejzaleski) [reply NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs]: Changed code to use `All` extension method

File: NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs (hunk 3 above)
Comment (maciejzaleski) [reply NTumbleBit/Proof/PoupardStern/Implementation/PoupardStern.cs]: Changed to used `Enumerable.Range`

**Comment by NicolasDorier** [comment]
@osagga  can you quick check

**Comment by NicolasDorier** [comment]
Make sense. I just wait @osagga go to be sure. But this seems perfect to me.

**Pull request merged**